src/content/blog/2024-12-04-differences-between-linting-and-type-checking.md

```diff
@@ -0,0 +1,248 @@
+---
+layout: post
+title: "Differences between linting and type checking"
+teaser: "Linters such as ESLint and type checkers such as TypeScript catch different areas of code defects — and are best used in conjunction with each other."
+authors:
+  - joshuakgoldberg
+categories:
+  - Storytime
+tags:
+  - Linting
+---
+
+Linters and type checkers are two kinds of tools that both analyze code and report on detected issues.
+They may seem similar at first.
+However, the two kinds of tools detect very different issues and are useful in different ways.
+
+## Definitions
+
+*"Static analysis"* is the term for tooling that checks code for issues without running it.
+There are three forms of static analysis commonly used in web development projects today:
+
+1. **Formatters**: only print your code quickly, without worrying about logic
+2. **Linters**: execute individually configurable checks known as "lint rules"
+3. **Type checkers**: collect all files into a full understanding of the project
+
+We'll focus in this blog post on *linters*, namely ESLint, and *type checkers*, namely [TypeScript](https://typescriptlang.org).
+ESLint and TypeScript are often used together to detect defects in code.
+
+## Differences in Purpose
+
+Type checkers make sure the *intent* behind values in code matches the *usage* of those values.
+They check that code is "type-safe": meaning values are used according to the ways their types describe as being allowed.
+
+Type checkers do not attempt to look for defects that are only likely, rather than certain.
+Nor do type checkers enforce subjective opinions that might change between projects.
+That means TypeScript does not attempt to enforce general best practices — only that values are used in type-safe ways.
+TypeScript won't detect likely mistakes that work within those types.
+
+Linters, on the other hand, primarily target *likely* defects and can also be used to enforce subjective opinions.
+ESLint and other linters catch issues that are technically type-safe but are potential sources of bugs.
+Many developers rely on linters to make sure their code follow framework and language best practices.
+
+For example, developers sometimes leave out the `break` or `return` at the end of a `switch` statement's `case`.
+Doing so is type-safe and permitted by JavaScript and TypeScript.
+In practice, this is almost always a mistake that allows the next `case` statement to run accidentally.
+ESLint's [`no-fallthrough`](https://eslint.org/docs/latest/rules/no-fallthrough) can catch that likely mistake:
+
+```ts
+function logFruit(value: "apple" | "banana" | "cherry") {
+    switch (value) {
+        case "apple":
+            console.log("🍏");
+            break;
+
+        case "banana":
+            console.log("🍌");
+
+        // eslint(no-fallthrough):
+        // Expected a 'break' statement before 'case'.
```

Comment:
I don't _love_ this first example:

* It's pretty verbose
* It doesn't show something that's _clearly_, _definitely_ a valuable report _("what if I want to break through, that's too opinionated!")_

...but I couldn't find a better lint rule that:

* Produces clear valuable report on a likely bug
* Is in `js.configs.recommended`
* Isn't superseded by TypeScript

Seeking inputs on a better example to show!

Reply:
`no-unused-vars`? Although I guess TypeScript is now flagging that, too.

`no-unexpected-multiline`?

Reply:
Yeah, and `no-unexpected-multiline` is pretty much made unnecessary by formatters... _sigh_